### PR TITLE
arch: common: semihosting: file operations

### PR DESCRIPTION
**Pull request by JordanYates**
Implements the ARM semihosting mechanism for performing file operations on the host system when a device is emulated or running under a debugger. The RISC-V architecture also implements semihosting using ARM's API.

One use case for this feature is when attempting to validate code on datasets larger than those that can fit in the ROM of an emulated image. With this feature, arbitrarily large files can simply be read from the host filesystem.

The file operations are not implemented as standard library functions (_open, _lseek, etc) as the arguments and behaviours of the ARM mechanism do not map cleanly to the standard posix functions. 

For example the `open` call returns a file handle (like posix open) but takes a non-standard integer for the file mode, which corresponds to strings passed to the posix fopen call.

This work also extends `CONFIG_CONSOLE_SEMIHOST` to work on all emulated ARM architectures, not just Cortex-M3.

More information on semihosting can be found here:
https://developer.arm.com/documentation/dui0471/m/what-is-semihosting-
Register and exception call semantics can be found here:
https://github.com/ARM-software/abi-aa/blob/main/semihosting/semihosting.rst

### CODE REVIEW

**Comment by nashif** [comment]
shouldn't this go into its own subsys instead of a generic library/os component that is usually reserved for cross-arch features?

File: include/sys/semihost.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+int semihost_flen(int fd);
+
+/**
+ * @brief Seeks to an absolute position in a file.
+ *
+ * @param fd handle returned by @ref semihost_open.
+ * @param offset offset from the start of the file in bytes.
+ *
+ * @retval 0 on success.
+ * @retval negative error code on failure.
```
Comment (gmarull) [inline include/sys/semihost.h:93]: nit: format is `@retval VALUE DESCRIPTION`, so maybe use `@retval -errno Negative error...`

Comment (de-nordic) [reply include/sys/semihost.h:93]: Would that be host `errno` codes or Zephyr `errno` codes?

**Comment by JordanYates** [comment]
> shouldn't this go into its own subsys instead of a generic library/os component that is usually reserved for cross-arch features?

Ideally the function implementations belong in a file common to `ARM` and `ARM64`, with the supervisor call done in arch specific code. Making this a generic subsys doesn't make much sense IMO because any other architectures are probably going to have their own bespoke API's of what they support.

**Comment by JordanYates** [comment]
Implementation has been moved to be under `arch`.
Semihosting is now supported for all ARM targets, not just Cortex-M3

File: cmake/emu/qemu.cmake [hunk 2]
```diff
@@ -70,7 +70,7 @@ endif()
 list(APPEND QEMU_FLAGS -serial chardev:con)
 
 # Connect semihosting console to the console chardev if configured.
-if(CONFIG_SEMIHOST_CONSOLE)
+if(CONFIG_ARM_SEMIHOST)
```
Comment (nordic-krch) [inline cmake/emu/qemu.cmake:73]: is this change needed?

File: include/arch/common/arm_semihost.h [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	ARM_SEMIHOST_OPEN_AB_PLUS       = 11,
+};
+
+/**
+ * @brief Manually execute a semihosting instruction
+ *
+ * @param instr instruction code to run
+ * @param args instruction specific arguments
+ *
+ * @return integer return code of instruction
```
Comment (nordic-krch) [inline include/arch/common/arm_semihost.h:105]: any more details what might be ok value and which is an error? Many functions bellow return directly after calling this function.

**Comment by microbuilder** [comment]
Can you add a sample application (doesn't need to be complex) demonstrating how this can be used, along with a short README point people to the API documentation, etc.

Ran the tests on a Cortex-M device (`$ ./scripts/twister -p mps2_an521 -N --inline-logs -T tests/arch/arm/arm_semihost`) and they seem to pass, but a sample and a bit of documentation would help people understand how to use this more effectively, and bring it to their attention.

**Comment by JordanYates** [comment]
> Can you add a sample application (doesn't need to be complex) demonstrating how this can be used, along with a short README pointing people to the API documentation, etc.

I can, but the sample cannot really do anything more than what the sample is performing (we have no knowledge of the install path of Zephyr or the build directory). So all that can really be done is make a file, write some data to it, and read it back out (what the test does).

File: cmake/emu/qemu.cmake (hunk 2 above)
Comment (JordanYates) [reply cmake/emu/qemu.cmake:73]: Yes, `-semihosting-config enable=on` is needed for semihosting to work

File: include/arch/common/arm_semihost.h (hunk 3 above)
Comment (JordanYates) [reply include/arch/common/arm_semihost.h:105]: Not really, this function is only exposed in case a user wants to run one of the other instructions. The meanings of the return codes varies by instruction, and would need to be  checked per instruction here: https://developer.arm.com/documentation/dui0471/m/what-is-semihosting-

**Comment by microbuilder** [comment]
> I can, but the sample cannot really do anything more than what the sample is performing (we have no knowledge of the install path of Zephyr or the build directory). So all that can really be done is make a file, write some data to it, and read it back out (what the test does).

Perhaps just add some documentation to describe the API, then, and the links in this PR so that it gets into the rendered documentation.

**Comment by JordanYates** [comment]
> Perhaps just add some documentation to describe the API, then, and the links in this PR so that it gets into the rendered documentation.

Done

**Comment by JordanYates** [comment]
With the merge of #44566 this has been extended to cover semihosting for ARM, ARM64 and RISC-V platforms.
Conveniently RISC-V semihosting is explicitly following the ARM API.
IMO this is enough to make the ARM implementation a "common" one.

**Review by gmarull — DISMISSED** [review]

File: drivers/console/Kconfig [hunk 4]
```diff
@@ -298,17 +298,12 @@ config NATIVE_POSIX_CONSOLE_INIT_PRIORITY
 config SEMIHOST_CONSOLE
 	bool "Use semihosting for console"
 	select CONSOLE_HAS_DRIVER
-	depends on CPU_CORTEX_M || ARM64 || RISCV
+	depends on ARM_SEMIHOST || RISCV
```
Comment (stephanosio) [inline drivers/console/Kconfig]: `ARM_SEMIHOST` is not a valid Kconfig symbol. Did you mean `SEMIHOST`?

File: tests/drivers/console/prj_semihost.conf [hunk 5]
```diff
@@ -2,4 +2,5 @@ CONFIG_TEST=y
 CONFIG_CONSOLE=y
 CONFIG_SERIAL=n
 CONFIG_UART_CONSOLE=n
+CONFIG_ARM_SEMIHOST=y
```
Comment (stephanosio) [inline tests/drivers/console/prj_semihost.conf]: same here

File: include/zephyr/arch/common/semihost.h [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+ * @brief Write the contents of a buffer into a file.
+ *
+ * @param fd handle returned by @ref semihost_open.
+ * @param buf buffer to write data from.
+ * @param len number of bytes to write.
+ *
+ * @retval 0 on success.
+ * @retval -errno negative error code on failure.
+ */
+long semihost_write(long fd, void *buf, long len);
```
Comment (stephanosio) [inline include/zephyr/arch/common/semihost.h:194]: `const void *buf`

File: arch/common/semihost.c [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+ * Organisation (CSIRO) ABN 41 687 119 230.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <zephyr/kernel.h>
+#include <zephyr/arch/common/semihost.h>
+
+struct _semihost_poll_in_args {
```
Comment (stephanosio) [inline arch/common/semihost.c:12]: Please do not add any new identifiers that start with `_` (i.e. reserved identifiers).

**Review by stephanosio — CHANGES_REQUESTED** [review]
Looks ok overall.

Just some minor comments.

**Review by stephanosio — APPROVED** [review]
LGTM

File: tests/arch/common/semihost/src/main.c [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <arch/common/semihost.h>
+
+static void test_file_ops(void)
+{
+	const char *test_file = "./test.bin";
+	uint8_t w_buffer[16] = { 1, 2, 3, 4, 5 };
```
Comment (bbolen) [inline tests/arch/common/semihost/src/main.c:14]: this could be made const

**Review by bbolen — APPROVED** [review]
